File: e2e/overview.test.js
// The overview's time range, and the empty state it used to get wrong.
//
// The bug worth pinning: the "nothing has reported yet" screen was gated on the
// count in the current window, so a project with months of history that was
// quiet overnight was told to install the SDK, and lost its tiles, its chart
// and its errors to say so.

const { test, before, after } = require('node:test')
const assert = require('node:assert')

const { BASE, launch, linger, signIn, createProject, seedLogs } = require('./harness')

let browser, page, busy, quiet, fresh

const ago = ms => new Date(Date.now() - ms)
const HOUR = 3600_000
const DAY = 24 * HOUR

async function openOverview(id, query = '') {
  await page.goto(`${BASE}/project/${id}/overview${query}`)
  await page.waitForSelector('h1')
}

before(async () => {
  ;({ browser, page } = await launch())
  await signIn(page)

  // Logged in the last hour: busy in every range.
  busy = await createProject(page, 'Overview Busy')
  await seedLogs(busy.id, busy.secret, [
    { message: 'App launched', level: 'info', at: ago(30 * 60_000) },
    { message: 'Checkout failed', level: 'error', at: ago(25 * 60_000) },
    { message: 'Retrying', level: 'warning', at: ago(20 * 60_000) },
  ])

  // Logged ten days ago and nothing since: empty at 24h, present at 30d, and
  // the case the old condition got wrong.
  quiet = await createProject(page, 'Overview Quiet')
  await seedLogs(quiet.id, quiet.secret, [
    { message: 'Ancient history', level: 'info', at: ago(10 * DAY) },
    { message: 'Also ancient', level: 'error', at: ago(10 * DAY + HOUR) },
  ])

  // Never logged anything.
  fresh = await createProject(page, 'Overview Fresh')
})

after(async () => {
  await linger(page)
  if (browser) await browser.close()
})

test('a project that has never reported is told how to start', async () => {
  await openOverview(fresh.id)

  const body = await page.locator('main').textContent()
  assert.match(body, /Nothing has reported yet/)
  assert.match(body, /Show me the snippet/)

  // No range control: there is nothing to range over yet.
  assert.equal(await page.locator('[data-range]').count(), 0,
    'a project with no data was offered a time range')
})

test('a quiet window is not mistaken for a project that never reported', async () => {
  // Ten day old logs, so the default 24 hour window is empty.
  await openOverview(quiet.id)

  const body = await page.locator('main').textContent()
  assert.ok(!body.includes('Nothing has reported yet'),
    'a project with logs was told it has never reported')
  assert.ok(!body.includes('Show me the snippet'),
    'a working project was sent to the SDK setup page')

  // It says the window is empty, and offers the way out.
  assert.match(body, /Nothing was logged in last 24 hours/)
  assert.match(body, /Try a longer range/)

  // The tiles survive, at zero.
  assert.ok(await page.locator('[data-stat="Total logs"]').count() > 0,
    'the tiles went away on a quiet window')
})

test('a longer range finds the logs the short one could not', async () => {
  await openOverview(quiet.id)
  const short = await page.locator('[data-stat="Total logs"]').textContent()
  assert.match(short, /\b0\b/, `expected zero in the 24h window, got: ${short}`)

  await Promise.all([
    page.waitForURL(/range=30d/),
    page.click('[data-range="30d"]'),
  ])

  const long = await page.locator('[data-stat="Total logs"]').textContent()
  assert.match(long, /\b2\b/, `expected both logs in the 30d window, got: ${long}`)

  const body = await page.locator('main').textContent()
  assert.ok(!body.includes('Nothing was logged in'),
    'the 30 day window still reported itself empty')
})

test('the range is in the URL, so it survives a reload and can be shared', async () => {
  await openOverview(busy.id, '?range=7d')

  assert.equal(await page.locator('[data-range="7d"]').getAttribute('aria-current'), 'page',
    'the range in the URL did not select its tab')
  assert.match(await page.locator('main').textContent(), /Activity · last 7 days/)

  // And a nonsense range falls back rather than erroring, because anyone can
  // edit a query string.
  await openOverview(busy.id, '?range=nonsense')
  assert.equal(await page.locator('[data-range="24h"]').getAttribute('aria-current'), 'page')
})

test('the labels follow the range', async () => {
  await openOverview(busy.id)
  let body = await page.locator('main').textContent()
  assert.match(body, /Activity · last 24 hours/)
  assert.match(body, /vs yesterday/)

  await Promise.all([page.waitForURL(/range=7d/), page.click('[data-range="7d"]')])
  body = await page.locator('main').textContent()
  assert.match(body, /vs the previous 7 days/)
  assert.ok(!body.includes('vs yesterday'),
    'a week-long window still claimed a day-over-day delta')
})

// Retention decides what is honest to offer: a window longer than we keep draws
// a cliff and then flat ground, and nothing on the screen says why.
test('a range past retention is not offered', async () => {
  await page.goto(`${BASE}/settings?tab=retention`)
  await page.waitForSelector('input[name="retention_days"]')
  await page.fill('input[name="retention_days"]', '7')
  await Promise.all([
    page.waitForResponse(r => r.url().includes('/settings/retention') && r.request().method() === 'POST'),
    page.click('form:has(input[name="retention_days"]) button[type="submit"]'),
  ])

  await openOverview(busy.id)
  assert.equal(await page.locator('[data-range="30d"]').count(), 0,
    'a 30 day range was offered on a 7 day retention')
  assert.ok(await page.locator('[data-range="7d"]').count() > 0,
    'the 7 day range was dropped on a 7 day retention')

  // Put it back, since every other file in this suite shares the instance.
  await page.goto(`${BASE}/settings?tab=retention`)
  await page.waitForSelector('input[name="retention_days"]')
  await page.fill('input[name="retention_days"]', '30')
  await Promise.all([
    page.waitForResponse(r => r.url().includes('/settings/retention') && r.request().method() === 'POST'),
    page.click('form:has(input[name="retention_days"]) button[type="submit"]'),
  ])
})

File: view/overview_test.go
package view

import (
	"strings"
	"testing"
	"time"

	"github.com/getcodescout/code_scout/internal/domain"
	"github.com/google/uuid"
)

func overviewFor(window domain.OverviewWindow, everLogged bool, logs int64) OverviewData {
	start := time.Date(2026, 8, 9, 0, 0, 0, 0, time.UTC)
	stats := &domain.ProjectOverview{
		Logs:            logs,
		EverLogged:      everLogged,
		Window:          window,
		WindowStartHour: start,
	}
	for i := 0; i < window.Columns(); i++ {
		stats.Buckets = append(stats.Buckets, domain.LogStatsBucket{
			Hour: start.Add(time.Duration(i) * window.Bucket),
		})
	}
	if logs > 0 {
		stats.Buckets[0].TotalCount = logs
	}
	return OverviewData{
		User:      &domain.User{Name: "T", Email: "t@test.local"},
		Project:   &domain.Project{},
		ProjectID: uuid.New(),
		Stats:     stats,
		Window:    window,
		Windows:   domain.OverviewWindowsFor(30),
		ShowDelta: window.ComparableAgainstPrevious(30),
	}
}

// The bug this screen had: the empty state was gated on the count in the
// window, so a project with a year of history that was quiet overnight was told
// it had never reported and lost its tiles, its chart and its errors to say so.
func TestAQuietWindowIsNotAProjectThatNeverReported(t *testing.T) {
	d := overviewFor(domain.OverviewWindowFor("7d", 30), true, 0)
	html := render(t, OverviewPage(d))

	if strings.Contains(html, "Nothing has reported yet") {
		t.Error("a project that has logged before was told it never has")
	}
	if strings.Contains(html, "Show me the snippet") {
		t.Error("a working project was sent to the SDK setup page")
	}
	// It still says the window was quiet, and offers the way out.
	if !strings.Contains(html, "Nothing was logged in last 7 days") {
		t.Error("the screen did not say the window itself was empty")
	}
	if !strings.Contains(html, "Try a longer range") {
		t.Error("no suggestion of what to do about an empty window")
	}
	// And the tiles are still there, at zero.
	if !strings.Contains(html, "Total logs") {
		t.Error("the tiles went away")
	}
}

// The genuine first run still gets the first-run screen, or the test above
// would pass with the fix removed.
func TestAProjectThatHasNeverReportedStillGetsTheSnippet(t *testing.T) {
	d := overviewFor(domain.DefaultOverviewWindow, false, 0)
	html := render(t, OverviewPage(d))

	if !strings.Contains(html, "Nothing has reported yet") {
		t.Error("a brand new project did not get the first-run state")
	}
	if !strings.Contains(html, "Show me the snippet") {
		t.Error("the first-run state lost its call to action")
	}
	// No range picker on a project with nothing to range over.
	if strings.Contains(html, `data-range=`) {
		t.Error("a project with no data was offered a time range")
	}
}

// The range is a link, so the view is shareable and the back button works.
func TestTheRangeIsALinkCarryingItsStateInTheURL(t *testing.T) {
	d := overviewFor(domain.OverviewWindowFor("7d", 30), true, 120)
	html := render(t, OverviewPage(d))

	for _, want := range []string{`data-range="24h"`, `data-range="7d"`, `data-range="30d"`, "?range=7d"} {
		if !strings.Contains(html, want) {
			t.Errorf("the range picker is missing %s", want)
		}
	}
	if !strings.Contains(html, `data-range="7d" aria-current="page"`) {
		t.Error("the active range is not marked current")
	}
}

// Retention decides what is honest to offer. A seven day instance must not
// offer a month, because the chart would draw a cliff and then flat ground and
// nothing on screen would say why.
func TestARangeRetentionCannotFillIsNotOffered(t *testing.T) {
	d := overviewFor(domain.DefaultOverviewWindow, true, 40)
	d.Windows = domain.OverviewWindowsFor(7)
	html := render(t, OverviewPage(d))

	if strings.Contains(html, `data-range="30d"`) {
		t.Error("a range past retention was offered")
	}
	if !strings.Contains(html, `data-range="7d"`) {
		t.Error("a range inside retention was dropped")
	}
}

// The labels move with the range. "vs yesterday" on a month of data is wrong,
// and so is "worst hour" when a column is a day.
func TestTheLabelsFollowTheRange(t *testing.T) {
	week := overviewFor(domain.OverviewWindowFor("7d", 30), true, 500)
	week.Stats.PeakErrorCount = 3
	week.Stats.PeakErrorHour = week.Stats.WindowStartHour
	html := render(t, OverviewPage(week))

	if !strings.Contains(html, "vs the previous 7 days") {
		t.Error("the delta still compares against yesterday on a week")
	}
	if strings.Contains(html, "vs yesterday") {
		t.Error("a week-long window claimed a day-over-day delta")
	}
	if !strings.Contains(html, "Activity · last 7 days") {
		t.Error("the chart heading is still hardcoded to 24 hours")
	}
	if !strings.Contains(html, "worst hour for errors") {
		t.Error("the callout lost its noun")
	}

	month := overviewFor(domain.OverviewWindowFor("30d", 60), true, 900)
	month.Stats.PeakErrorCount = 9
	month.Stats.PeakErrorHour = month.Stats.WindowStartHour
	if got := render(t, OverviewPage(month)); !strings.Contains(got, "worst day for errors") {
		t.Error("a daily bucket is still called an hour")
	}
}

// The window fitting inside retention is not enough: the delta reaches back
// another whole window. Printing a rise against a month that was deleted is
// worse than printing no delta at all.
func TestNoDeltaIsShownWhenThePreviousPeriodIsGone(t *testing.T) {
	d := overviewFor(domain.OverviewWindowFor("30d", 30), true, 5000)
	d.ShowDelta = false
	html := render(t, OverviewPage(d))

	if strings.Contains(html, "vs the previous 30 days") {
		t.Error("a delta was shown against a period retention has deleted")
	}
	// The counts themselves are still there.
	if !strings.Contains(html, "Total logs") {
		t.Error("hiding the delta took the tile with it")
	}
}
